src/popen.c

```diff
@@ -165,37 +165,59 @@ int mypclose(FILE *fp, pid_t pid) {
 
     errno = 0;
 
-    siginfo_t info;
-    if(waitid(P_PID, (id_t) pid, &info, WEXITED) != -1) {
-        switch(info.si_code) {
-            case CLD_EXITED:
-                if(info.si_status)
-                    error("child pid %d exited with code %d.", info.si_pid, info.si_status);
-                return(info.si_status);
-
-            case CLD_KILLED:
-                error("child pid %d killed by signal %d.", info.si_pid, info.si_status);
-                return(-1);
-
-            case CLD_DUMPED:
-                error("child pid %d core dumped by signal %d.", info.si_pid, info.si_status);
-                return(-2);
-
-            case CLD_STOPPED:
-                error("child pid %d stopped by signal %d.", info.si_pid, info.si_status);
-                return(0);
-
-            case CLD_TRAPPED:
-                error("child pid %d trapped by signal %d.", info.si_pid, info.si_status);
-                return(-4);
-
-            case CLD_CONTINUED:
-                error("child pid %d continued by signal %d.", info.si_pid, info.si_status);
-                return(0);
-
-            default:
-                error("child pid %d gave us a SIGCHLD with code %d and status %d.", info.si_pid, info.si_code, info.si_status);
-                return(-5);
+#ifdef __OpenBSD__
+    int options = 0;
+#else
+    int options = WEXITED;
+#endif
+    int proc_status;
+    if(waitpid(pid, &proc_status, options) > 0) {
```

Comment:
From what I can tell of the FreeBSD and OpenBSD man pages, this is an appropriate replacement. It's confusing though, so I might not be correct.

Reply:
hm... I would be more comfortable if you move this `waitpid()` code to a new static function and override the whole function for OpenBSD, so that it runs as it was in Linux and FreeBSD.

Reply:
In general, OpenBSD tries to avoid patching things by adding `#ifdef`s, because they make the code harder to debug and more error-prone. `waitpid()` is available on every platform, so if I can confirm that it behaves the same way, it would be a better replacement. After work today, I’ll pull both versions of this function out into a test program and test every code path to make sure they behave the same way.

Reply:
Check the source of `waitid()`: https://code.woboq.org/userspace/glibc/sysdeps/posix/waitid.c.html
It calls `waitpid()`, but does a few more things:

`options = WEXITED` may not work for `waitpid()`. The code removes it explicitly at the `waitpid()` call:

```c
child = __waitpid (pid, &status,
options
#ifdef WEXITED
&~ WEXITED
#endif
);
```

Then it uses many macros to set the result we need (more than `WIFSIGNALED`):

```c
if (WIFEXITED (status))
{
infop->si_code = CLD_EXITED;
infop->si_status = WEXITSTATUS (status);
}
else if (WIFSIGNALED (status))
{
infop->si_code = WCOREDUMP (status) ? CLD_DUMPED : CLD_KILLED;
infop->si_status = WTERMSIG (status);
}
else if (WIFSTOPPED (status))
{
infop->si_code = CLD_STOPPED;
infop->si_status = WSTOPSIG (status);
}
...
```